apps/prairielearn/assets/scripts/instructorFileEditorClient.ts

```diff
@@ -0,0 +1,193 @@
+  element: HTMLElement;
+  diskHash?: string;
+  saveElement?: HTMLButtonElement | null;
+  inputContentsElement?: HTMLInputElement | null;
+  editor: ace.Ace.Editor;
+
+  constructor({
+    element,
+    saveElement,
+    aceMode,
+    readOnly,
+    contents,
+  }: {
+    element: HTMLElement;
+    saveElement?: HTMLButtonElement | null;
+    aceMode?: string;
+    readOnly?: boolean;
+    contents?: string;
+  }) {
+    this.element = element;
+    this.diskHash = element.querySelector<HTMLInputElement>(
+      'input[name=file_edit_orig_hash]',
+    )?.value;
+    const editorElement = element.querySelector<HTMLElement>('.editor');
+    if (!editorElement) {
+      throw new Error(`Could not find .editor element inside ${element.id}`);
+    }
+
+    this.saveElement = saveElement;
+    this.inputContentsElement = element.querySelector('input[name=file_edit_contents]');
+    this.editor = ace.edit(editorElement, {
+      minLines: 10,
+      maxLines: Infinity,
+      autoScrollEditorIntoView: true,
+      wrap: true,
+      showPrintMargin: false,
+      mode: aceMode || undefined,
+      readOnly,
+      enableKeyboardAccessibility: true,
+      theme: 'ace/theme/chrome',
+    });
+
+    this.setEditorContents(contents ? this.b64DecodeUnicode(contents) : '');
+
+    this.editor.commands.addCommand({
+      name: 'saveAndSync',
+      bindKey: { win: 'Ctrl-s', mac: 'Command-s' },
+      exec: () => {
+        this.saveElement?.click();
+      },
+    });
+
+    this.editor.getSession().on('change', () => this.onChange());
+  }
+
+  setEditorContents(contents: string) {
+    // use session.setValue to reset the undo stack as well
+    this.editor.getSession().setValue(contents);
+    this.editor.gotoLine(1, 0, false);
+    this.editor.focus();
+    this.syncFileToHiddenInput();
+  }
+
+  syncFileToHiddenInput() {
+    if (this.inputContentsElement) {
+      this.inputContentsElement.value = this.b64EncodeUnicode(this.editor.getValue());
+    }
+  }
+
+  b64DecodeUnicode(str: string) {
+    // Going backwards: from bytestream, to percent-encoding, to original string.
+    return decodeURIComponent(
+      atob(str)
+        .split('')
+        .map((c) => {
+          return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
+        })
+        .join(''),
+    );
+  }
+
+  b64EncodeUnicode(str: string) {
+    // first we use encodeURIComponent to get percent-encoded UTF-8,
+    // then we convert the percent encodings into raw bytes which
+    // can be fed into btoa.
+    return btoa(
+      encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, (_match, p1) => {
+        return String.fromCharCode(Number(`0x${p1}`));
+      }),
+    );
+  }
+
+  onChange() {
+    this.syncFileToHiddenInput();
+    this.checkDiff();
+  }
+
+  getHash(contents: string) {
+    return CryptoJS.SHA256(this.b64EncodeUnicode(contents)).toString();
+  }
+
+  checkDiff() {
+    if (this.saveElement) {
+      const curHash = this.getHash(this.editor.getValue());
```

Comment:
I know this didn't change with your code, but as long as we're changing this: is there a reason we don't use a simple string comparison of the current and original values? It's not a huge deal, but it'd be nice to drop a dependency. On the server, we can use Node's native `createHash('sha256').update(...).digest(...)`.

Reply:
Note that the editor's initial content is not necessarily the disk contents (e.g., when there is a conflicting change), so this may imply duplicating the disk content. Also the router code also uses the same hash comparison to identify changes. I'd prefer to coordinate this change across client- and server-side code in a separate PR, to make sure this information is properly processed, likely the same that would strengthen the types of the parameters passed between router and html.ts (see other comment).